### PR TITLE
Suggest let for possible binding with ty

### PR DESCRIPTION
**Pull request by chenyukang**
Origin from https://github.com/rust-lang/rust/pull/109128#discussion_r1179866137

r? @Nilstrieb

### CODE REVIEW

File: compiler/rustc_parse/src/parser/diagnostics.rs [hunk 1]
```diff
@@ -399,6 +399,22 @@ impl<'a> Parser<'a> {
                 }
             }
         }
+        // we suggest add the missing `let` before the identifier
+        // `a: Ty = 1` -> `let a: Ty = 1`
+        if self.token == token::Colon {
+            let prev_span = self.prev_token.span.shrink_to_lo();
+            let snapshot = self.create_snapshot_for_diagnostic();
+            self.bump();
+            if self.parse_ty().is_ok() && self.token == token::Eq {
```
Comment (Nilstrieb) [inline compiler/rustc_parse/src/parser/diagnostics.rs:408]: You need to cancel the error here. This ICEs:
```rust
struct A {
    : :u8,
}
```

**Review by Noratrieb — CHANGES_REQUESTED** [review]
I don't think this is a great place to put it (the error message is very specialized for statements, but the location is pretty generic), I'd prefer it if you found a more specific place. After fixing your ICE locally, this snippet produces the following output:
```rust
struct A {
    : u8 =,
}
```
```
error: expected identifier, found `:`
 --> uwu.rs:6:5
  |
5 | struct A {
  |        - while parsing this struct
6 |     : u8 =,
  |     ^ expected identifier
  |
help: you might have meant to introduce a new binding
  |
5 | struct A let {
  |          +++
```
which, while this snippet is totally nonsensical, is not very good.

File: compiler/rustc_parse/src/parser/diagnostics.rs [hunk 2]
```diff
@@ -1006,6 +1006,35 @@ impl<'a> Parser<'a> {
...
+        if self.token == token::Colon {
+            let prev_span = self.prev_token.span.shrink_to_lo();
+            let snapshot = self.create_snapshot_for_diagnostic();
+            self.bump();
+            match self.parse_ty() {
+                Ok(_) => {
+                    if self.token == token::Eq {
+                        err.span_suggestion_verbose(
+                            prev_span,
+                            "you might have meant to introduce a new binding",
```
Comment (Nilstrieb) [inline compiler/rustc_parse/src/parser/diagnostics.rs:1024]: could you turn this into a session diagnostic derive?

**Review by Noratrieb — COMMENTED** [review]
That implementation is a lot nicer! r=me after addressing the comment

**Pull request merged**